File: finances-nextjs-gin-postgresql/backend/e2e/insert_transaction..go
package e2e

import (
	"api/transactions"
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func InsertTransaction(t *testing.T, token string, categoryID, typeTransactioName, description string, amount float64) {
	reqBody := transactions.InsertTransactionParams{
		Amount:              amount,
		TypeTransactionName: typeTransactioName,
		CategoryID:          categoryID,
		Description:         description,
	}
	reqBodyJson, _ := json.Marshal(reqBody)

	// Enviar uma requisição real para o servidor rodando
	client := &http.Client{Timeout: 10 * time.Second}
	req, err := http.NewRequest(
		"POST",
		fmt.Sprintf("%s/transactions", serverEndPoint),
		bytes.NewBuffer(reqBodyJson),
	)
	assert.Nil(t, err, "expected error nil")

	// Adicionar cabeçalhos, como o token de autenticação ou outros
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+token)

	resp, err := client.Do(req)
	assert.Nil(t, err, "expected error nil")
	assert.NotNil(t, resp, "expected resp")
	defer resp.Body.Close()

	// Verificar o status code
	assert.Equal(t, resp.StatusCode, http.StatusCreated)

	// Verificar o corpo da resposta
	var responseBody map[string]interface{}
	err = json.NewDecoder(resp.Body).Decode(&responseBody)
	assert.Nil(t, err, "expected error nil")

	// Validar os dados retornados
	assert.Equal(t, "transação realizada", responseBody["message"])
}

File: finances-nextjs-gin-postgresql/frontend/services/insert-transaction.ts
import { ServiceResult } from "./service-result"

export type InsertTransactionParams = {
  amount: number,
  typeTransactionName: string,
  categoryId: string,
  description: string,
}

const url = `${process.env.NEXT_PUBLIC_ENDPOINT_API}/transactions`

export const insertTransaction = (token: string) =>
  async (params: InsertTransactionParams): Promise<ServiceResult<void>> => {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Authorization": `Bearer ${token}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify(params)
    })

    const resultContent = await response.json()

    return {
      error: !response.ok,
      message: resultContent.message as string
    }
  }